File: scripts/delete_e_line_products.dart
import 'dart:io';
import 'package:firebase_core/firebase_core.dart';
import 'package:firebase_database/firebase_database.dart';
import '../lib/firebase_options.dart';

Future<void> deleteELineProducts() async {
  print('🔍 Initializing Firebase...');
  
  // Initialize Firebase
  await Firebase.initializeApp(
    options: DefaultFirebaseOptions.currentPlatform,
  );
  
  final database = FirebaseDatabase.instance;
  
  try {
    print('📊 Fetching all products from database...');
    
    // Get all products
    final snapshot = await database.ref('products').once();
    
    if (!snapshot.snapshot.exists || snapshot.snapshot.value == null) {
      print('❌ No products found in database');
      return;
    }
    
    final productsData = snapshot.snapshot.value as Map<dynamic, dynamic>;
    print('✅ Found ${productsData.length} total products\n');
    
    // Find E line products
    final eLineProducts = <String, Map<dynamic, dynamic>>{};
    
    productsData.forEach((key, value) {
      if (value is Map) {
        final sku = value['sku']?.toString() ?? '';
        final model = value['model']?.toString() ?? '';
        final name = value['name']?.toString() ?? '';
        final productType = value['productType']?.toString() ?? '';
        
        // Check if it's an E line product
        // E line products typically have SKUs ending with -E or contain E- in the model
        if (sku.endsWith('-E') || 
            sku.contains('E-') ||
            model.endsWith('-E') || 
            model.contains('E-') ||
            productType.toLowerCase().contains('e series') ||
            productType.toLowerCase().contains('e-series') ||
            name.toLowerCase().contains('e series') ||
            name.toLowerCase().contains('e-series')) {
          eLineProducts[key.toString()] = value;
        }
      }
    });
    
    if (eLineProducts.isEmpty) {
      print('✅ No E line products found in database');
      return;
    }
    
    print('🔍 Found ${eLineProducts.length} E line products:\n');
    
    // List E line products
    int index = 1;
    for (final entry in eLineProducts.entries) {
      final product = entry.value;
      print('$index. SKU: ${product['sku']} | Model: ${product['model']} | Name: ${product['name']}');
      index++;
    }
    
    print('\n⚠️  WARNING: This will permanently delete ${eLineProducts.length} E line products!');
    print('Type "DELETE" to confirm deletion, or press Enter to cancel:');
    
    final confirmation = await readInput();
    
    if (confirmation.toUpperCase() != 'DELETE') {
      print('❌ Deletion cancelled');
      return;
    }
    
    print('\n🗑️  Deleting E line products...');
    
    // Delete each E line product
    int deleted = 0;
    int failed = 0;
    
    for (final key in eLineProducts.keys) {
      try {
        await database.ref('products/$key').remove();
        deleted++;
        print('  ✅ Deleted product with key: $key');
      } catch (e) {
        failed++;
        print('  ❌ Failed to delete product with key: $key - Error: $e');
      }
    }
    
    print('\n📊 Deletion Summary:');
    print('  ✅ Successfully deleted: $deleted products');
    if (failed > 0) {
      print('  ❌ Failed to delete: $failed products');
    }
    
    // Verify remaining products
    final verifySnapshot = await database.ref('products').once();
    if (verifySnapshot.snapshot.exists && verifySnapshot.snapshot.value != null) {
      final remainingProducts = verifySnapshot.snapshot.value as Map;
      print('\n📊 Database now contains ${remainingProducts.length} products');
      
      // Double-check no E line products remain
      int remainingELine = 0;
      remainingProducts.forEach((key, value) {
        if (value is Map) {
          final sku = value['sku']?.toString() ?? '';
          final model = value['model']?.toString() ?? '';
          if (sku.endsWith('-E') || sku.contains('E-') || 
              model.endsWith('-E') || model.contains('E-')) {
            remainingELine++;
          }
        }
      });
      
      if (remainingELine > 0) {
        print('  ⚠️  Warning: $remainingELine E line products may still remain');
      } else {
        print('  ✅ All E line products successfully removed');
      }
    }
    
  } catch (e) {
    print('❌ Error: $e');
  }
}

Future<String> readInput() async {
  final input = stdin.readLineSync() ?? '';
  return input.trim();
}

void main() async {
  await deleteELineProducts();
  print('\n✅ Script completed');
  exit(0);
}

File: scripts/delete_e_line_products.py
import firebase_admin
from firebase_admin import credentials, db
import json
import os
import sys

# Set UTF-8 encoding for Windows console
if sys.platform == 'win32':
    sys.stdout.reconfigure(encoding='utf-8')

# Initialize Firebase Admin SDK
# Check if service account key exists
if os.path.exists('../taquotes-firebase-adminsdk-fbsvc-ae05f60f37.json'):
    cred = credentials.Certificate('../taquotes-firebase-adminsdk-fbsvc-ae05f60f37.json')
elif os.path.exists('taquotes-firebase-adminsdk-fbsvc-ae05f60f37.json'):
    cred = credentials.Certificate('taquotes-firebase-adminsdk-fbsvc-ae05f60f37.json')
elif os.path.exists('../firebase-admin-key.json'):
    cred = credentials.Certificate('../firebase-admin-key.json')
elif os.path.exists('firebase-admin-key.json'):
    cred = credentials.Certificate('firebase-admin-key.json')
else:
    print('❌ Error: Firebase admin key not found')
    print('Please download the service account key from Firebase Console:')
    print('1. Go to Firebase Console > Project Settings > Service Accounts')
    print('2. Click "Generate new private key"')
    print('3. Save the file as firebase-admin-key.json in the project root')
    exit(1)

firebase_admin.initialize_app(cred, {
    'databaseURL': 'https://taquotes-default-rtdb.firebaseio.com'
})

def delete_e_line_products():
    print('🔍 Fetching all products from database...')
    
    try:
        # Get reference to products
        ref = db.reference('products')
        products_data = ref.get()
        
        if not products_data:
            print('❌ No products found in database')
            return
        
        print(f'✅ Found {len(products_data)} total products\n')
        
        # Find E line products
        e_line_products = {}
        
        for key, product in products_data.items():
            sku = product.get('sku', '')
            model = product.get('model', '')
            name = product.get('name', '')
            product_type = product.get('productType', '')
            
            # Check if it's an E line product
            if (sku.endswith('-E') or 
                'E-' in sku or
                model.endswith('-E') or 
                'E-' in model or
                'e series' in product_type.lower() or
                'e-series' in product_type.lower() or
                'e series' in name.lower() or
                'e-series' in name.lower()):
                e_line_products[key] = product
        
        if not e_line_products:
            print('✅ No E line products found in database')
            return
        
        print(f'🔍 Found {len(e_line_products)} E line products:\n')
        
        # List E line products
        for i, (key, product) in enumerate(e_line_products.items(), 1):
            print(f"{i}. SKU: {product.get('sku')} | Model: {product.get('model')} | Name: {product.get('name')}")
        
        print(f'\n⚠️  WARNING: This will permanently delete {len(e_line_products)} E line products!')
        confirmation = input('Type "DELETE" to confirm deletion, or press Enter to cancel: ').strip()
        
        if confirmation.upper() != 'DELETE':
            print('❌ Deletion cancelled')
            return
        
        print('\n🗑️  Deleting E line products...')
        
        # Delete each E line product
        deleted = 0
        failed = 0
        
        for key in e_line_products.keys():
            try:
                db.reference(f'products/{key}').delete()
                deleted += 1
                print(f'  ✅ Deleted product with key: {key}')
            except Exception as e:
                failed += 1
                print(f'  ❌ Failed to delete product with key: {key} - Error: {e}')
        
        print('\n📊 Deletion Summary:')
        print(f'  ✅ Successfully deleted: {deleted} products')
        if failed > 0:
            print(f'  ❌ Failed to delete: {failed} products')
        
        # Verify remaining products
        remaining_products = db.reference('products').get() or {}
        print(f'\n📊 Database now contains {len(remaining_products)} products')
        
        # Double-check no E line products remain
        remaining_e_line = 0
        for product in remaining_products.values():
            sku = product.get('sku', '')
            model = product.get('model', '')
            if (sku.endswith('-E') or 'E-' in sku or 
                model.endswith('-E') or 'E-' in model):
                remaining_e_line += 1
        
        if remaining_e_line > 0:
            print(f'  ⚠️  Warning: {remaining_e_line} E line products may still remain')
        else:
            print('  ✅ All E line products successfully removed')
        
    except Exception as e:
        print(f'❌ Error: {e}')

if __name__ == '__main__':
    delete_e_line_products()
    print('\n✅ Script completed')